python/tools/REx_calibrate_trajectory_limits.py

```diff
@@ -0,0 +1,1918 @@
+) -> str:
+    """
+    We've landed on a bad step. This means that we've exceeded the calibration targets.
+    We now want to set this as the upper bound, and bring the good step as close as possible.
+    """
+
+    is_first_time_backtracking = calibration_data.last_bad_calibration is None
+    calibration_data.last_bad_calibration = motion_data
+
+    motion_data.step_calibration_result = (
+        StepCalibrationResult.BACKTRACKING_INCREASING_TIME
+    )
+
+    if is_first_time_backtracking:
+        motion_data.step_calibration_result = StepCalibrationResult.FIRST_OVERSHOOT
+
+    # Add the motion_data after adding all the correct statuses and attributes.
+    calibration_data.motion_data.append(motion_data)
+
+    backtracking_message= ""
+    if run_mode == _RunMode.dynamic_limit_mode:
+        backtracking_message = 'Step_calibration will backtrack to find the optimal calibration values.'
+
+    return motion_data.motion_overview(
+        joint=joint,
+        prefix=f"""
+{calibration_data.direction_name} {joint.name} {calibration_data.profile_name} Motion Profile: 
+The  dynamic range has been exceeded. {backtracking_message}
+""",
+    )
+
+
+def _get_next_travel_duration_in_seconds(
+    calibration_data: TrajectoryCalibrationData,
+) -> tuple[float, float, str]:
+    """
+    This method figures our how much the next calibration step/run is going to change the travel duration by.
+
+    Returns new_travel_duration, change_time_by and message
+    """
+    new_travel_duration = (
+        calibration_data.calibration_targets.travel_duration_start_seconds
+    )
+    change_time_by = new_travel_duration
+
+    good_travel_duration = (
+        calibration_data.last_good_calibration.trajectory.travel_duration_seconds
+        if calibration_data.last_good_calibration
+        else None
+    )
+    bad_travel_duration = (
+        calibration_data.last_bad_calibration.trajectory.travel_duration_seconds
+        if calibration_data.last_bad_calibration
+        else None
+    )
+    STOPPED_FOR_SAFETY_MOTION_FIXED_DECREASE = (
+        calibration_data.calibration_targets.end_condition_time_step
+    )
+
+    if not good_travel_duration and not bad_travel_duration:
+        return new_travel_duration, change_time_by, ""
+
+    message = f"""New {calibration_data.direction_name} calibration step:"""
+
+    if (
+        good_travel_duration
+        and not bad_travel_duration
+        and calibration_data.last_good_calibration
+    ):
+        # We need to speed up motion and decrease the time step:
+        change_time_by = -1 * _get_dynamic_decrease_time_by(
+            calibration_data=calibration_data,
+            motion_data=calibration_data.last_good_calibration,
+        )
+
+        new_travel_duration = round(good_travel_duration + change_time_by, 2)
+
+        message += f"""
+    Decreasing the last travel time from {good_travel_duration}s to {new_travel_duration}s ({change_time_by}s) for this run.
```

Comment:
You can use `({change_time_by:.2f}s)` to limit change_time_by to print with 2 sig figs since new_travel_duration is rounded anyways. This eliminates printouts like:

```
Decreasing the last travel time from 6.88s to 5.02s (-1.8559750000000002s) for this run.
```